### PR TITLE
Report api and trigger updates

### PR DESCRIPTION
**Pull request by shrikantdange1208**

### CODE REVIEW

File: functions/api/action.js [hunk 1]
```diff
@@ -213,7 +213,7 @@ function validateParams(body, type) {
                 branch: joi.string().min(1).max(30).required(),
                 product: joi.string().min(1).max(30).required(),
                 productName: joi.string().min(1).max(30).required(),
-                operationalQuantity: joi.number().required(),
+                operationalQuantity: joi.number().integer().strict().integer().strict().required(),
```
Comment (neerk28) [inline functions/api/action.js]: conditions are twice.Can you remove one

File: functions/api/branch.js [hunk 2]
```diff
@@ -6,8 +6,10 @@ const formatDate = require('../common/dateFormatter')
 const joi = require('@hapi/joi');
 const admin = require('firebase-admin');
 const { isAdminOrSuperAdmin, isSuperAdmin } = require('../middleware/auth');
+const functions = require('firebase-functions')
+const { isAdmin } = require('../middleware/auth');
```
Comment (neerk28) [inline functions/api/branch.js]: there is no "isAdmin". Pls. use "isAdminOrSuperAdmin" instead

File: functions/api/branch.js [hunk 3]
```diff
@@ -254,18 +259,51 @@ function validateParams(body, type) {
...
+async function deleteThresholdsFromAllProducts(branchId) {
+    console.log(`Deleting thresholds for branch ${branchId} from all products`)
+    const productCollecton = db.collection(constants.PRODUCTS)
+    const productSnapshots = await productCollecton.get()
+    productSnapshots.forEach(async product => {
+        const productData = product.data()
+        if(productData[constants.THRESHOLDS].hasOwnProperty(branchId)) {
+            console.log(`Deleting threshold for branch ${branchId} from product ${productData[constants.NAME]}`)
+            delete productData[constants.THRESHOLDS][branchId]
+            await product.ref.update({'thresholds': productData[constants.THRESHOLDS]}, { merge: true })
```
Comment (neerk28) [inline functions/api/branch.js:305]: i somehow feel, this is an indefinite loop. where you update product and each product keeps coming again in foreach loop. Please double check once.

File: functions/api/product.js [hunk 4]
```diff
@@ -8,6 +8,7 @@ const joi = require('@hapi/joi')
 const admin = require('firebase-admin')
 const functions = require('firebase-functions')
 const express = require('express');
+const c = require('config');
```
Comment (neerk28) [inline functions/api/product.js]: is it being used?

File: functions/api/report.js [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+ * @param {*} product 
+ * @param {*} fromDate 
+ * @param {*} toDate 
+ */
+async function getReportForProduct(branchDocument, product, fromDate, toDate) {
+
+    let productTransactionRef = await branchDocument
+        .collection(constants.TRANSACTIONS)
+        .where(constants.PRODUCT, '==', product[constants.PRODUCT])
+        .where(constants.DATE, '>', fromDate)
```
Comment (neerk28) [inline functions/api/report.js:120]: not including fromdate and todate ?

**Review by neerk28 — CHANGES_REQUESTED** [review]
Havent looked into reports in detail. Would you great if you can go over the flow to understand better.

File: functions/api/action.js (hunk 1 above)
Comment (shrikantdange1208) [reply functions/api/action.js]: Done

File: functions/api/branch.js (hunk 2 above)
Comment (shrikantdange1208) [reply functions/api/branch.js]: removed

File: functions/api/branch.js (hunk 3 above)
Comment (shrikantdange1208) [reply functions/api/branch.js:305]: It actually does not. We can sync up and I can explain the flow

File: functions/api/product.js (hunk 4 above)
Comment (shrikantdange1208) [reply functions/api/product.js]: removed

File: functions/api/report.js (hunk 5 above)
Comment (shrikantdange1208) [reply functions/api/report.js:120]: I am. in this method fromDate is fromDate -1 and toDate is toDate + 1

**Review by neerk28 — APPROVED** [review]
LGTM

File: functions/api/report.js (hunk 5 above)
Comment (ramyarv) [reply functions/api/report.js:120]: @shrikantdange1208 isnt fromDate meant to be inclusive. so in that case 2020-07-18 => >= 2020-07-18 00:00:00

File: functions/common/constants.js [hunk 6]
```diff
@@ -64,6 +64,10 @@ define("TRANSACTIONS", "transactions");
 define("PENDING_REQUESTS", "pendingRequests");
 define("NEXT_PAGE_TOKEN","nextPageToken");
 define("PREV_PAGE_TOKEN","prevPageToken");
+define("PAGE_SIZE",3)
```
Comment (ramyarv) [inline functions/common/constants.js]: page_size is already available.

**Review by ramyarv — COMMENTED** [review]
minor comments

File: functions/api/report.js (hunk 5 above)
Comment (shrikantdange1208) [reply functions/api/report.js:120]: This fromDate is fromDate - 1 and toDate is toDate + 1

**Pull request merged**